core/module_specifier.rs

```diff
@@ -49,17 +109,50 @@ impl ModuleSpecifier {
   /// example is allowed here, whereas in import statements a leading "./" is
   /// required ("./foo.js"). This function is aware of the current working
   /// directory and returns an absolute URL.
-  pub fn resolve_root(
-    root_specifier: &str,
-  ) -> Result<ModuleSpecifier, url::ParseError> {
-    if let Ok(url) = Url::parse(root_specifier) {
-      Ok(ModuleSpecifier(url))
-    } else {
-      let cwd = std::env::current_dir().unwrap();
-      let base = Url::from_directory_path(cwd).unwrap();
-      let url = base.join(root_specifier)?;
-      Ok(ModuleSpecifier(url))
+  pub fn resolve_from_cwd(
+    specifier: &str,
+  ) -> Result<ModuleSpecifier, ResolveError> {
+    if let Ok(module_specifier) = ModuleSpecifier::resolve_absolute(specifier) {
+      return Ok(module_specifier);
+    }
+
+    // fallback to relative file path
+    // TODO: factor out as `normalize_path` method - it is used in /cli/deno_dir.rs and should
+    //  be applied to resolve https://github.com/denoland/deno/issues/1798
+    // HACK: `Url::from_directory_path` is used here because it normalizes the path.
+    // Joining `/dev/deno/" with "./tests" using `PathBuf` yields `/deno/dev/./tests/`.
+    // On the other hand joining `/dev/deno/" with "./tests" using `Url` yields "/dev/deno/tests"
+    // - and that's what we want.
+    // There exists similar method on `PathBuf` - `PathBuf.canonicalize`, but the problem
+    // is `canonicalize` resolves symlinks and we don't want that.
+    // We just want o normalize the path...
+    let specifier_path = PathBuf::from(specifier);
+    let cwd = std::env::current_dir().unwrap();
+    let path = cwd.join(specifier_path);
+    ensure_valid_prefix(&path)?;
```

Comment:
Since you're processing *relative* paths here, I don't think you need to validate that it has a prefix. The cases we should support are:
`/foo/bar`  <-- not entirely absolute on windows but close enough. You can check for it with the `has_root()` method.
`c:/foo/bar`  <-- that's where the prefix matters. We want either a drive letter + `has_root() == true`, or nothing at all.
'a/relative/path`  <-- for relative paths there shouldn't be any prefix. If there is then we get `C:a/relative/path` and that's *not* what we want to support.

Reply:
@piscisaureus I don't have access ATM to Windows machine, could you help me out with a patch?

Reply:
@bartlomieju Funny thing, I don't have one either ATM, because the battery has blown up.
The good news is that tomorrow my new laptop should be delivered, so yes I will help you then.

Reply:
Great, while you're at it, would you mind taking a look at `test_bad_prefix()`?